### DISCUSSION TITLE
ServerBuilder is never dropped

### DISCUSSION THREAD

**Discussion by jwiesler** [actix/actix-web#2428] [Q&A]
Hi,

I noticed that `ServerBuilder` is not dropped after the server is stopped. It doesn't matter whether `ServerBuilder::exit` is true and `System::current().stop()` is called (it really should happen then shouldn't it?) or not. Furthermore, the `ServerBuilder` [future ](https://github.com/actix/actix-net/blob/master/actix-server/src/builder.rs#L431) can never complete because it returns `Pending` even when the command receiver returns `None`!

Is this intended? I'm relying on `web::Data<T>` being dropped after the server shuts down, but `ServerBuilder` holds the last reference to it forever.

**Comment by jwiesler** [comment]
`ServerBuilder` is dropped after `main` exits. This leaves two questions:
* If `ServerBuilder` runs on the same `System` as `main`, what does `System::stop` even do? I can await futures in main even after `stop` was called.
* How do I fix this? How can I run more cleanup code after `ServerBuilder` dropped? Would running the actix sytem for the server from a tokio main method help?

**Comment by fakeshadow** [comment]
https://github.com/actix/actix-net/blob/1c8fcaebbc2f385acece09f7fcc2fd565daae1c4/actix-server/tests/test_server.rs#L19

You need another scope to do anything after runtime shutdown. If you just straight exit main when runtime shutdown you can't do any meaningful work.

That said in general you are encouraged to do all the work inside the actix-runtime context. Design your clean up logic to fit it could be a better solution overall.

**Comment by jwiesler** [comment]
> That said in general you are encouraged to do all the work inside the actix-runtime context. 

Are there any examples for this? I need to shut down background workers but I know of no way to react to the server shutting down or even receive a notification that it does so.

And how should I handle shutting down IO? In theory it might take ages to write the last 10 bytes of an operation and stopping the System will interrupt the write process and corrupt the file.
